File: ABC184/E.py
import sys
import heapq

H, W = map(int, sys.stdin.readline().strip().split())
inf = 10 ** 5

table = []
for h in range(H):
    table.append(list(str(sys.stdin.readline().strip())))

warp = {chr(ord("a") + i): [] for i in range(26)}
for h in range(H):
    for w in range(W):
        if table[h][w].isalpha():
            if table[h][w].islower():
                warp[table[h][w]].append((h, w))

            elif table[h][w] == 'S':
                start = (0, h, w)
            else:
                goal = (h, w)

que = [start]
heapq.heapify(que)
dist = [[inf] * W for i in range(H)]
dist[start[1]][start[2]] = 0

while que:
    du, hu, wu = heapq.heappop(que)

    # ゴール確認
    if (hu, wu) == goal:
        print(du)
        sys.exit()

    # テレポート
    if table[hu][wu].isalpha() and table[hu][wu].islower():
        if warp[table[hu][wu]]:
            for hv, wv in warp[table[hu][wu]]:
                if dist[hv][wv] == inf:
                    dist[hv][wv] = du + 1
                    heapq.heappush(que, (du + 1, hv, wv))

    for x, y in [[0, 1], [0, -1], [1, 0], [-1, 0]]:
        hv = hu + x
        wv = wu + y

        if (hv < 0) or (hv >= H) or (wv < 0) or (wv >= W):
            continue

        if table[hv][wv] == "#":
            continue

        if dist[hv][wv] != inf:
            continue

        heapq.heappush(que, (du + 1, hv, wv))
        dist[hv][wv] = du + 1

print(-1)

File: ABC184/E.cpp
#include<bits/stdc++.h>
using namespace std;

const int INF = 1001001;

int main() {
    int H, W;
    cin >> H >> W;

    string s;
    vector<string> table;
    for (int i = 0; i < H; i++) {
        cin >> s;
        table.push_back(s);
    }

    // vector<pair<int, int>> warp(26);
    map<char, vector<pair<int, int>>> warp;
    pair<int, int> start, goal;
    for (int h = 0; h < H; h++) {
        for (int w = 0; w < W; w++) {
            if (isalpha(table[h][w]) && islower(table[h][w])) {
                warp[table[h][w]].push_back(make_pair(h, w));
            } else if (table[h][w] == 'S') start = make_pair(h, w);
            else if (table[h][w] == 'G') goal = make_pair(h, w);
        }
    }

    queue<pair<int, int>> q;
    q.emplace(start);
    vector<pair<int, int>> d = {{-1, 0}, {1, 0}, {0, 1}, {0, -1}};
    vector<vector<int>> dist(H, vector<int>(W, INF));
    // vector dist(H, vector(W, INF));
    dist[start.first][start.second] = 0;

    while(!q.empty()) {

        int hu = q.front().first;
        int wu = q.front().second;
        q.pop();

        if (hu == goal.first && wu == goal.second) {
            cout << dist[hu][wu] << endl;
            return 0;
        }


        for (auto du :  d) {
            int hv = hu + du.first;
            int wv = wu + du.second;

            if (hv < 0 || hv >= H || wv < 0 || wv >= W) continue;
            if (table[hv][wv] == '#') continue;

            if (dist[hv][wv] != INF) continue;

            dist[hv][wv] = dist[hu][wu] + 1;
            q.emplace(hv, wv);
        }
        
        // warp 処理
        if (isalpha(table[hu][wu]) && islower(table[hu][wu])) {
            for (auto [h, w] : warp[table[hu][wu]]) {
                if (dist[h][w] != INF) continue;
                dist[h][w] = dist[hu][wu] + 1;
                q.emplace(h, w);
            }
            warp[table[hu][wu]] = {};
        }


    }
    cout << -1 << endl;

}